doc/design/python_api.md

```diff
@@ -0,0 +1,186 @@
+
+The implementation of the first step is a temporary implementation. We should design our Python API concepts based on `compile-time` concepts. We just use `runtime` classes to implement it for now.
+
+
+## Python Class and compile-time protobuf
+
+Since we design our Python API concepts based on `compile-time`, we try to map our Python classes to every compile-time result, i.e., the protobuf messages. They are:
+
+
+| Python Class | Compile-time protobuf |
+| --- | --- |
+| Program | ProgramDesc |
+| Block | BlockDesc |
+| Operator | OpDesc |
+| Variable | VarDesc |
+
+### Program
+
+`Program` is the description of the whole training process and there can only be one `Program` object, which is created automatically by the system at the very beginning. `Program` is formed by a series of `Block`.
+
+```python
+class Program(objects):
+    def __init__(self):
+        self.blocks = vector<Block>()
+        self.blocks.append(Block(None))
+        self.current_block_idx = 0
+
+    def get_block(block_idx):
+        return self.blocks[block_idx]
+
+    def current_block():
+        return self.get_block(self.current_block_idx)
+    
+    def fallback_current_block():
+        self.current_block_idx = self.current_block().parent_idx
+
+
+    def create_block():
+        new_block_idx = len(self.block)
+        self.blocks.append(Block(parent_idx=self.current_block_idx,
+                                 idx=new_block_idx))
+        self.current_block_idx = new_block_idx
+```
+
+`Program` will create the first block in its constructor. The first block is called 'global block'. It is where all parameters are stored.
+
+### Block
+
+Block is just like programming languages `{}`, which contains many operators and variables. There are two data fields in `Block`.  1) An associate map, whose key is variable name and value is variable itself; 2) A list of operators.
+
+The block is hierarchical because PaddlePaddle supports RNN and IfElse. For example, RNN is like `for-loop` in programming languages. There is new `block` inside a `for-loop`. To represent hierarchies, `Block` stores the index of `parent Block` inside. The 'index' means the block's position in `Program`'s `blocks`. If `parent_idx=None`, the block itself is the outermost block, i.e., the 'global block'.
+
+
+```python
+class Block(objects):
+    def __init__(self, parent_idx, idx):
+        self.vars = map<string, Variable>()
+        self.ops = vector<Operator>()
+        self.idx = idx
+        self.parent_idx = parent_idx
+    
+    def create_var(self, ...):
+        # create variable in `self.vars`
+        return Variable(...)
+    
+    
+    def create_global_var(self, ...):
+        if self.parent_idx is not None:
+            parent_block = program.get_block(parent_idx)
+            return parent_block.create_global_var(...)
+        else:
+            return self.create_var(...)
+    
+    def create_parameter(self, ...):
+        return self.create_global_var(...)
```

Comment:
Does `create_parameter` mean `create_global_var`?

I guess yes, because:

> All parameters should be stored in the global block

What happens if the one block's `create_parameter` call creates a parameter that has the same name from another block? Intuitively block A and block B creating the parameter with the same name will not interfere with each other.

Reply:
Actually, no.

1. All parameters must be created in the global block.
2. All parameters are marked as Parameter. The Parameter is a subclass of Variable.

Reply:
So the user can not create parameter inside a if-else block? (e.g., create a FC layer with parameter W and B).